From 0d37b338d2cc9a54db438a2f542d06e4478fb20c Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@gmail.com>
Date: Mon, 27 Oct 2025 09:47:05 -0400
Subject: [PATCH] boot: conditionally emit rdinit warning to avoid spurious
 messages

The rdinit warning added by commit 98aa4d5d242d3 ("init/main.c: add
warning when file specified in rdinit is inaccessible") is helpful for
debugging initramfs issues, but generates spurious warnings on systems
that don't use initramfs or have rdinit explicitly set.

Only emit the warning when:
 - An initramfs is actually available (initrd_start || phys_initrd_size), OR
 - rdinit was explicitly set on the command line

This preserves the diagnostic value for legitimate initramfs boot failures
while eliminating noise on traditional root filesystem boots.

Signed-off-by: Bruce Ashfield <bruce.ashfield@gmail.com>
---
 init/main.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/init/main.c b/init/main.c
index d9cf20a20f919..dfddfa6c8a489 100644
--- a/init/main.c
+++ b/init/main.c
@@ -161,6 +161,7 @@ static size_t initargs_offs;
 
 static char *execute_command;
 static char *ramdisk_execute_command = "/init";
+static bool rdinit_cmdline_set __initdata;
 
 /*
  * Used to generate warnings if static_key manipulation functions are used
@@ -622,6 +623,7 @@ static int __init rdinit_setup(char *str)
 	unsigned int i;
 
 	ramdisk_execute_command = str;
+	rdinit_cmdline_set = true;
 	/* See "auto" comment in init_setup */
 	for (i = 1; i < MAX_INIT_ARGS; i++)
 		argv_init[i] = NULL;
@@ -1575,6 +1577,11 @@ void __init console_on_rootfs(void)
 
 static noinline void __init kernel_init_freeable(void)
 {
+#ifdef CONFIG_BLK_DEV_INITRD
+	bool initrd_available;
+#else
+	bool initrd_available = false;
+#endif
 	/* Now the scheduler is fully set up and can do blocking allocations */
 	gfp_allowed_mask = __GFP_BITS_MASK;
 
@@ -1606,6 +1613,14 @@ static noinline void __init kernel_init_freeable(void)
 
 	kunit_run_all_tests();
 
+#ifdef CONFIG_BLK_DEV_INITRD
+	/*
+	 * Capture whether an initrd was supplied before wait_for_initramfs()
+	 * resets the initrd markers.
+	 */
+	initrd_available = initrd_start || phys_initrd_size;
+#endif
+
 	wait_for_initramfs();
 	console_on_rootfs();
 
@@ -1616,8 +1631,10 @@ static noinline void __init kernel_init_freeable(void)
 	int ramdisk_command_access;
 	ramdisk_command_access = init_eaccess(ramdisk_execute_command);
 	if (ramdisk_command_access != 0) {
-		pr_warn("check access for rdinit=%s failed: %i, ignoring\n",
-			ramdisk_execute_command, ramdisk_command_access);
+		if (initrd_available || rdinit_cmdline_set) {
+			pr_warn("check access for rdinit=%s failed: %i, ignoring\n",
+				ramdisk_execute_command, ramdisk_command_access);
+		}
 		ramdisk_execute_command = NULL;
 		prepare_namespace();
 	}
-- 
2.39.2

